### PR TITLE
Ports - Margaret 

### PR DESCRIPTION
**Pull request by msfinnan**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | forEach loops, if/else if/else statements, && / || logic
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Yes! Instead of using the Ruby Docs, I have been using MDN. I used StackOverflow for both. 
What was a challenge you were able to overcome on this assignment? | Learning how to do basic debugging. I didn't know how to do that when I started the project, and then got help to use console.log and it made debugging a lot easier. Also learned that you can't break out of a forEach loop. 
What has been interesting and positive about learning a new programming language? | Getting a better understanding of what is programming and what is Ruby specific. 
What is something to focus on your learnings in JavaScript in the next week? | More about debugging, get more comfortable using classes.

### CODE REVIEW

**Comment by tildeee** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | x
**Code Requirements** |
`drawLetters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `drawLetters` pass | x
`usesAvailableLetters` method | x
All tests for `usesAvailableLetters` pass | x
`scoreWord` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `scoreWord` pass | x
`highestScoreFrom` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highestScoreFrom` pass | x
**Overall** |



Well done on this project, Margaret! Your solutions and code are all spot-on and you definitely got the solution I was looking for. Your project and code is readable and clean!


I have a series of comments that are definitely just small suggestions for improving your code style, and also making your code more JavaScript-y.


Overall, great work!

File: src/adagrams.js [hunk 1]
```diff
@@ -1,7 +1,97 @@
+const _ = require('lodash');
+
 const Adagrams = {
   drawLetters() {
-    // Implement this method for wave 1
+    const allLetters = ["A", "A", "A", "A", "A", "A", "A", "A", "A", "B", "B", "C", "C", "D", "D", "D", "D", "E", "E", "E", "E", "E", "E", "E", "E", "E", "E", "E", "E", "F", "F", "G", "G", "G", "H", "H", "I", "I", "I", "I", "I", "I", "I", "I", "I", "J", "K", "L", "L", "L", "L", "M", "M", "N", "N", "N", "N", "N", "N", "O", "O", "O", "O", "O", "O", "O", "O", "P", "P", "Q", "R", "R", "R", "R", "R", "R", "S", "S", "S", "S", "T", "T", "T", "T", "T", "T", "U", "U", "U", "U", "V", "V", "W", "W", "X", "Y", "Y", "Z"];
+
+    //use lodash .sampleSize method to get a random sample of 10 letters 
+    const lettersInHand = _.sampleSize(allLetters, 10);
```
Comment (tildeee) [inline src/adagrams.js:8]: NICE USING LODASH! hehehe :-)

File: src/adagrams.js [hunk 2]
```diff
@@ -1,7 +1,97 @@
...
+
+    return lettersInHand;
+
+  },
+
+  usesAvailableLetters(input, lettersInHand) {
+    input = input.toUpperCase();
+
+    let lettersObject = {};
```
Comment (tildeee) [inline src/adagrams.js:17]: turns out that `lettersObject` can be a `const` because you never re-assign it, even if you modify it

File: src/adagrams.js [hunk 3]
```diff
@@ -1,7 +1,97 @@
...
+
+    lettersInHand.forEach(function (letter) {
+      if (lettersObject[letter] !== undefined) {
+        lettersObject[letter] += 1;
+      } else {
+        lettersObject[letter] = 1;
+      }
+    });
+
+    const inputArray = input.split("");
```
Comment (tildeee) [inline src/adagrams.js:27]: Hm, it felt weird that above you basically re-assigned `input` as itself with `input = input.toUpperCase()`. And now here, you make `inputArray`. It might feel better if you inline all of it into one call: `const inputArray = input.toUpperCase().split('');`. Also, I tend to prefer variable names that talk less about the data type and more about meaning, so maybe consider something like `letters` here instead. Lastly: Don't forget our preference for single-quotes ;)

File: src/adagrams.js [hunk 4]
```diff
@@ -1,7 +1,97 @@
...
+
+    let boolean = true;
```
Comment (tildeee) [inline src/adagrams.js:29]: Hmm, I wish there were a better name for this variable besides `boolean`...

File: src/adagrams.js [hunk 5]
```diff
@@ -1,7 +1,97 @@
...
+
+  scoreWord(word) {
+    word = word.toUpperCase()
+
+    const letterScores = {
+      'A': 1, 'B': 3, 'C': 3, 'D': 2, 'E': 1, 'F': 4, 'G': 2, 'H': 4,
+      'I': 1, 'J': 8, 'K': 5, 'L': 1, 'M': 3, 'N': 1, 'O': 1, 'P': 3,
+      'Q': 10, 'R': 1, 'S': 1, 'T': 1, 'U': 1, 'V': 4, 'W': 4, 'X': 8,
+      'Y': 4, 'Z': 10
+    };
```
Comment (tildeee) [inline src/adagrams.js:51]: Turns out JS defaults all keys to strings, so you can use this syntax instead: `const letterScores = { A: 1, B: 3, C: 3, ...}` etc

File: src/adagrams.js [hunk 6]
```diff
@@ -1,7 +1,97 @@
...
+
+    let score = 0;
+
+    const wordArray = word.split("");
```
Comment (tildeee) [inline src/adagrams.js:55]: As I mentioned above, it may make sense to consider refactoring to `const wordArray = word.toUpperCase().split('');`. Also, for the variable name, I may consider changing the name to `letters` here too

File: src/adagrams.js [hunk 7]
```diff
@@ -1,7 +1,97 @@
...
+
+    wordArray.forEach(function (letter) {
+      score += letterScores[letter]
```
Comment (tildeee) [inline src/adagrams.js:58]: don't forget your semicolon!

File: src/adagrams.js [hunk 8]
```diff
@@ -1,7 +1,97 @@
...
+    });
```
Comment (tildeee) [inline src/adagrams.js:59]: Well done and this is correct implementation! For refactoring, consider using JS's reduce function ;) https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

File: src/adagrams.js [hunk 9]
```diff
@@ -1,7 +1,97 @@
...
+
+  highestScoreFrom(words) {
+
+    const winningWord = {
+      word: "",
+      score: 0,
+    };
+
+    words.forEach(function (word) {
+      if (Adagrams.scoreWord(word) > winningWord['score']) {
```
Comment (tildeee) [inline src/adagrams.js:76]: instead of invoking `Adagrams.scoreWord(word)` five times in this method, consider storing the value of the score in a variable up top! Honestly though, efficiency at this scale isn't critical, so this refactoring isn't crucial

File: src/adagrams.js [hunk 10]
```diff
@@ -1,7 +1,97 @@
...
+        winningWord['word'] = word;
+        winningWord['score'] = Adagrams.scoreWord(word);
+      } else if (Adagrams.scoreWord(word) === winningWord['score']) {
+        if (word.length === 10 && winningWord['word'].length !== 10) {
+          winningWord['word'] = word;
+          winningWord['score'] = Adagrams.scoreWord(word);
+        } else if (word.length < winningWord['word'].length && winningWord['word'].length !== 10) {
+          winningWord['word'] = word;
+          winningWord['score'] = Adagrams.scoreWord(word);
```
Comment (tildeee) [inline src/adagrams.js:85]: I'm not really very concerned about this code style, but using dot notation for the `winningWord` object is technically less characters than bracket notation (aka if you used `winningWord.word` and `winningWord.score`) hahaha
